File: src/main/java/twists/worldless/WorldlessCommand.java
package twists.worldless;

import com.mojang.brigadier.CommandDispatcher;
import net.minecraft.command.CommandRegistryAccess;
import net.minecraft.command.argument.TimeArgumentType;
import net.minecraft.server.command.CommandManager;
import net.minecraft.server.command.ServerCommandSource;

import static net.minecraft.server.command.CommandManager.*;

/// I wish I wrote this in Kotlin
public class WorldlessCommand {

    public static void register(CommandDispatcher<ServerCommandSource> dispatcher, CommandRegistryAccess registryAccess, CommandManager.RegistrationEnvironment environment) {
        dispatcher.register(literal("twist").then(literal("worldless").then(
                literal("start").executes(context -> {
                    if (context.getSource().getServer() instanceof WorldlessStateHolder holder) {
                        WorldlessState worldlessState = holder.twists$worldless$getWorldlessState();
                        worldlessState.setEnabled(true);
                    }
                    return 1;
                }))
                .then(literal("stop").executes(context -> {
                    if (context.getSource().getServer() instanceof WorldlessStateHolder holder) {
                        WorldlessState worldlessState = holder.twists$worldless$getWorldlessState();
                        worldlessState.setEnabled(false);
                    }
                    return 1;
                }))
                .then(literal("pause").executes(context -> {
                    if (context.getSource().getServer() instanceof WorldlessStateHolder holder) {
                        WorldlessState worldlessState = holder.twists$worldless$getWorldlessState();
                        worldlessState.setPaused(true);
                    }
                    return 1;
                }))
                .then(literal("resume").executes(context -> {
                    if (context.getSource().getServer() instanceof WorldlessStateHolder holder) {
                        WorldlessState worldlessState = holder.twists$worldless$getWorldlessState();
                        worldlessState.setPaused(false);
                    }
                    return 1;
                }))
                .then(
                        literal("time")
                                .then(literal("add").then(argument("time", TimeArgumentType.time()).executes(context -> {
                                    if (context.getSource().getServer() instanceof WorldlessStateHolder holder) {
                                        long time = context.getArgument("time", Integer.class);
                                        WorldlessState worldlessState = holder.twists$worldless$getWorldlessState();
                                        worldlessState.modifyTicksUntilReset(time);

                                    }
                                    return 1;
                                }))
                                )
                                .then(literal("remove").then(argument("time", TimeArgumentType.time()).executes(context -> {
                                    if (context.getSource().getServer() instanceof WorldlessStateHolder holder) {
                                        long time = context.getArgument("time", Integer.class);
                                        WorldlessState worldlessState = holder.twists$worldless$getWorldlessState();
                                        worldlessState.modifyTicksUntilReset(-time);

                                    }
                                    return 1;
                                })))
                                .then(literal("set").then(argument("time", TimeArgumentType.time()).executes(context -> {
                                            if (context.getSource().getServer() instanceof WorldlessStateHolder holder) {
                                                long time = context.getArgument("time", Integer.class);
                                                WorldlessState worldlessState = holder.twists$worldless$getWorldlessState();
                                                worldlessState.setTicksUntilReset(time);

                                            }
                                            return 1;
                                        })))
                                .then(literal("setdefault").then(argument("time", TimeArgumentType.time()).executes(context -> {
                                    if (context.getSource().getServer() instanceof WorldlessStateHolder holder) {
                                        long time = context.getArgument("time", Integer.class);
                                        WorldlessState worldlessState = holder.twists$worldless$getWorldlessState();
                                        worldlessState.setTicksPerWorld(time);

                                    }
                                    return 1;
                                })))
                )

        ));

    }
}

File: src/main/kotlin/twists/command/WorldlessCommand.kt
package twists.command

import com.mojang.brigadier.builder.LiteralArgumentBuilder
import net.casual.arcade.commands.CommandTree
import net.casual.arcade.commands.argument
import net.casual.arcade.commands.literal
import net.casual.arcade.commands.requiresPermission
import net.minecraft.commands.CommandSourceStack
import net.minecraft.commands.arguments.TimeArgument
import twists.Twists

import twists.worldless.WorldlessMinecraftServerExtension.Companion.worldlessState

object WorldlessCommand {


    fun createWorldlessCommand(): LiteralArgumentBuilder<CommandSourceStack> {
        return CommandTree.buildLiteral("worldless") {
            requiresPermission(2)
            literal("start") {
                executes { it -> it.source.server.worldlessState.enabled = true; 1 }
            }
            literal("stop") {
                executes { it -> it.source.server.worldlessState.enabled = false; 1 }
            }
            literal("pause") {
                executes { it -> it.source.server.worldlessState.paused = true; 1 }
            }
            literal("resume") {
                executes { it -> it.source.server.worldlessState.paused = false; 1 }
            }

            literal("time") {
                literal("add") {
                    argument("time", TimeArgument.time()) {
                        executes { it ->
                            val time = it.getArgument("time", Int::class.java)
                            it.source.server.worldlessState.modifyTicksUntilReset(time.toLong())
                            1
                        }
                    }
                }
                literal("remove") {
                    argument("time", TimeArgument.time()) {
                        executes { it ->
                            val time = it.getArgument("time", Int::class.java)
                            it.source.server.worldlessState.modifyTicksUntilReset(-time.toLong())
                            1
                        }
                    }
                }
                literal("set") {
                    argument("time", TimeArgument.time()) {
                        executes { it ->
                            val time = it.getArgument("time", Int::class.java)
                            it.source.server.worldlessState.setTicksUntilReset(time.toLong())
                            1
                        }
                    }
                }
                literal("setdefault") {
                    argument("time", TimeArgument.time()) {
                        executes { it ->
                            val time = it.getArgument("time", Int::class.java)
                            it.source.server.worldlessState.setTicksPerWorld(time.toLong())
                            1
                        }
                    }
                }
            }

        }
    }


}